waku/v2/waku_relay/protocol.nim

```diff
@@ -131,31 +196,37 @@ iterator subscribedTopics*(w: WakuRelay): lent PubsubTopic =
 proc subscribe*(w: WakuRelay, pubsubTopic: PubsubTopic, handler: WakuRelayHandler) =
   debug "subscribe", pubsubTopic=pubsubTopic
 
-  var subsHandler: PubsubRawHandler
-  when handler is SubscriptionHandler:
-    subsHandler = proc(pubsubTopic: PubsubTopic, data: seq[byte]): Future[void] {.gcsafe.} =
-        let decodeRes = WakuMessage.decode(data)
-        if decodeRes.isErr():
-          debug "message decode failure", pubsubTopic=pubsubTopic, error=decodeRes.error
-          return
+  # we need to wrap the handler since gossipsub doesnt understand WakuMessage
+  let wrappedHandler = proc(pubsubTopic: string, data: seq[byte]): Future[void] {.gcsafe, raises: [Defect].} =
+    let decMsg = WakuMessage.decode(data)
+    if decMsg.isErr():
+      # fine if triggerSelf enabled, since validators are bypassed
+      error "failed to decode WakuMessage, validator passed a wrong message"
+      let fut = newFuture[void]()
+      fut.complete()
+      return fut
+    else:
+      return handler(pubsubTopic, decMsg.get())
 
-        handler(pubsubTopic, decodeRes.value)
-  else:
-    subsHandler = handler
+  # add the default validator to the topic
+  procCall GossipSub(w).addValidator(pubSubTopic, validator)
 
-  procCall GossipSub(w).subscribe(pubsubTopic, subsHandler)
+  # set this topic parameters for scoring
+  w.topicParams[pubsubTopic] = TopicParameters
 
-proc unsubscribe*(w: WakuRelay, topics: seq[TopicPair]) =
-  debug "unsubscribe", pubsubTopic=topics.mapIt(it[0])
+  # subscribe to the topic with our wrapped handler
+  procCall GossipSub(w).subscribe(pubsubTopic, wrappedHandler)
 
-  procCall GossipSub(w).unsubscribe(topics)
+proc unsubscribe*(w: WakuRelay, topics: PubsubTopic) =
+  debug "unsubscribe", pubsubTopic=topics.mapIt(it)
+
+  procCall GossipSub(w).unsubscribe(topics, nil)
 
 proc unsubscribeAll*(w: WakuRelay, pubsubTopic: PubsubTopic) =
   debug "unsubscribeAll", pubsubTopic=pubsubTopic
 
   procCall GossipSub(w).unsubscribeAll(pubsubTopic)
 
-
 proc publish*(w: WakuRelay, pubsubTopic: PubsubTopic, message: WakuMessage|seq[byte]): Future[int] {.async.} =
```

Comment:
If the idea is not to return a `seq[byte]` only `WakuMessage`s in the subscription handler anymore, then to be consistent, you should not allow publishing a `seq[byte]`, only `WakuMessage`s.

Reply:
yup, done dc70df97f859a4fa907db0be13bfd56d7a23dce8